src/main/java/org/commcare/cases/util/StorageBackedTreeRoot.java

```diff
@@ -291,23 +299,43 @@ protected void initBasicQueryHandlers(QueryPlanner queryPlanner) {
         }
 
 
-        String[] namesToMatch = new String[numKeysToProcess];
-        String[] valuesToMatch = new String[numKeysToProcess];
+        Vector<String> namesToMatch = new Vector<>();
+        Vector<String> valuesToMatch = new Vector<>();
+        Vector<String> namesToInverseMatch = new Vector<>();
+        Vector<String> valuesToInverseMatch = new Vector<>();
 
         String cacheKey = "";
         String keyDescription ="";
 
         for (int i = numKeysToProcess - 1; i >= 0; i--) {
-            namesToMatch[i] = profiles.elementAt(i).getKey();
-            valuesToMatch[i] = (String)
-                    (((IndexedValueLookup)profiles.elementAt(i)).value);
-
-            cacheKey += "|" + namesToMatch[i] + "=" + valuesToMatch[i];
-            keyDescription += namesToMatch[i] + "|";
+            String name;
+            String value;
+            String operator;
+            if (profiles.elementAt(i) instanceof IndexedValueLookup) {
+                name = profiles.elementAt(i).getKey();
+                value = (String)(((IndexedValueLookup)profiles.elementAt(i)).value);
+                namesToMatch.add(name);
+                valuesToMatch.add(value);
+                operator = "=";
+            } else {
+                name = profiles.elementAt(i).getKey();
+                value = (String)(((NegativeIndexedValueLookup)profiles.elementAt(i)).value);
+                namesToInverseMatch.add(name);
+                valuesToInverseMatch.add(value);
+                operator = "!=";                
+            }
+            cacheKey += "|" + name + operator + value;
+            keyDescription += name + "|";
         }
-        mMostRecentBatchFetch = new String[2][];
-        mMostRecentBatchFetch[0] = namesToMatch;
-        mMostRecentBatchFetch[1] = valuesToMatch;
+        String[] namesArray = namesToMatch.toArray(new String[namesToMatch.size()]);
+        String[] valuesArray = valuesToMatch.toArray(new String[valuesToMatch.size()]);
+        String[] inverseNames = namesToInverseMatch.toArray(new String[namesToInverseMatch.size()]);
+        String[] inverseValues = valuesToInverseMatch.toArray(new String[valuesToInverseMatch.size()]);
+        mMostRecentBatchFetch = new String[4][];
+        mMostRecentBatchFetch[0] = namesArray;
+        mMostRecentBatchFetch[1] = valuesArray;
+        mMostRecentBatchFetch[2] = inverseNames;
+        mMostRecentBatchFetch[3] = inverseValues;
```

Comment:
Bump on this one again too. It's fine if you don't have time to jump into it, but wanted to flag just in case it was possible.

Reply:
I thought I replied here, but must have actually sent it over slack. I am hesitant to jump into this because this object is only used in the mobile code (the only use in formplayer code is to clear it, it is never read), and I don't feel like I have enough context to make the correct changes, since I do not really know how it is used there.